portability-spi-cloud/src/main/java/org/datatransferproject/spi/cloud/storage/JobStoreWithValidator.java

```diff
@@ -0,0 +1,134 @@
+  }
+
+  @Override
+  public void updateJobAuthStateToCredsAvailable(UUID jobId) throws IOException {
+    PortabilityJob job = findJob(jobId);
+    // Set update job auth data
+    JobAuthorization jobAuthorization =
+        job.jobAuthorization().toBuilder().setState(CREDS_AVAILABLE).build();
+    job = job.toBuilder().setAndValidateJobAuthorization(jobAuthorization).build();
+    updateJob(
+        jobId,
+        job,
+        (previous, updated) -> validateForUpdateStateToCredsAvailable(previous));
+  }
+
+  private static void validateForUpdateStateToCredsAvailable(PortabilityJob job) {
+    String dataType = job.transferDataType();
+    Preconditions.checkArgument(
+        !Strings.isNullOrEmpty(dataType), "Missing valid dataTypeParam: %s", dataType);
+
+    String exportService = job.exportService();
+    Preconditions.checkArgument(
+        !Strings.isNullOrEmpty(exportService), "Missing valid exportService: %s", exportService);
+
+    String importService = job.importService();
+    Preconditions.checkArgument(
+        !Strings.isNullOrEmpty(importService), "Missing valid importService: %s", importService);
+    Preconditions.checkState(job.jobAuthorization().state() == JobAuthorization.State.INITIAL);
+  }
+
+  @Override
+  public void updateJobWithCredentials(UUID jobId, PortabilityJob job) throws IOException {
+    updateJob(
+        jobId,
+        job,
+        ((previous, updated) -> {
+          Preconditions.checkState(
+              previous.jobAuthorization().state() == CREDS_ENCRYPTION_KEY_GENERATED);
+          Preconditions.checkState(updated.jobAuthorization().state() == CREDS_STORED);
+        }));
+  }
+
+  @Override
+  public void markJobAsFinished(UUID jobId, State state) throws IOException {
+    Preconditions.checkState(state == State.ERROR || state == State.COMPLETE);
+    updateJobState(jobId, state, State.IN_PROGRESS, JobAuthorization.State.CREDS_STORED);
+  }
+
+  @Override
+  public void markJobAsStarted(UUID jobId) throws IOException {
+    updateJobState(jobId, State.IN_PROGRESS, State.NEW, JobAuthorization.State.CREDS_STORED);
+  }
+
+  @Override
+  public void markJobAsTimedOut(UUID jobId) throws IOException {
+    PortabilityJob job = findJob(jobId);
+    updateJob(
+        jobId,
+        job.toBuilder()
+            .setState(PortabilityJob.State.ERROR)
+            .setAndValidateJobAuthorization(
+                job.jobAuthorization()
+                    .toBuilder()
+                    .setState(JobAuthorization.State.TIMED_OUT)
+                    .build())
+            .build());
+  }
+
+  /**
+   * Verifies a {@code PortabilityJob} already exists for {@code jobId}, and updates the entry to
+   * {@code job}.
+   *
+   * @throws IOException if a job didn't already exist for {@code jobId} or there was a problem
+   * updating it
+   */
+  protected abstract void updateJob(UUID jobId, PortabilityJob job) throws IOException;
```

Comment:
Should this be in the base JobStore or JobStore with validator?

Reply:
I moved this to JobStoreWithValidator so there would be no generic update job method without validation on the JobStore. This then pushes every call to the JobStore to be a well defined state transition.